kernel/sched.c

```diff
@@ -665,24 +674,99 @@ void z_sched_start(struct k_thread *thread)
 	z_reschedule(&sched_spinlock, key);
 }
 
+/**
+ * @brief Halt a thread
+ *
+ * If the target thread is running on another CPU, flag it as needing to
+ * abort and send an IPI (if supported) to force a schedule point and wait
+ * until the target thread is switched out (ISRs will spin to wait and threads
+ * will block to wait). If the target thread is not running on another CPU,
+ * then it is safe to act immediately.
+ *
+ * @param thread Thread to suspend or abort
+ * @param key Current key for sched_spinlock
+ * @param terminate True if aborting thread, false if suspending thread
+ */
+static void z_thread_halt(struct k_thread *thread, k_spinlock_key_t key,
+			  bool terminate)
+{
+#ifdef CONFIG_SMP
+	if (is_halting(_current) && arch_is_in_isr()) {
```

Comment:
Finally found some time to comb through.

This is a race in the current version: checking is_halting() here outside the lock doesn't prevent another CPU from setting it during the code below before the lock is taken, and indeed that can deadlock because now you're spinning on another thread while your current thread is flagged aborting, which is a forbidden state.  The old version was inside the lock.

Reply:
@andyross - Assuming the lock to which you are referring is spinlock **_sched_lock_**, then I am not seeing what you are seeing. The routine **_z_thread_halt()_** is only called from two places--both of which obtain the lock on **_sched_lock_** prior to the routine **_z_thread_halt()_** being called. We are therefore inside the lock when we perform the check on _**is_halting()**_. And as far as I can tell, the only places that modify either of the **_THREAD_HALTING** bits also have spinlock **_sched_lock_** locked.

If there are still concerns, perhaps we could schedule a time to discuss this (and of course summarise the results here).

Reply:
@andyross - ping